File: backend/app/routers/segments.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional
from uuid import UUID

from ..database import get_db
from ..middleware.auth import get_current_user
from ..middleware.rbac import require_permission
from ..schemas.segment import SegmentCreate, SegmentUpdate, SegmentResponse
from ..services.segment_service import SegmentService
from ..models.user import User

router = APIRouter(prefix="/segments", tags=["segments"])

@router.get("", response_model=List[SegmentResponse], dependencies=[Depends(require_permission("segments:read"))])
async def list_segments(
    status: Optional[str] = "active",
    search: Optional[str] = None,
    limit: int = Query(20, le=100),
    cursor: Optional[UUID] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return await SegmentService.list_segments(db, status, search, limit, cursor)

@router.post("", response_model=SegmentResponse, status_code=status.HTTP_201_CREATED, dependencies=[Depends(require_permission("segments:create"))])
async def create_segment(
    data: SegmentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return await SegmentService.create_segment(db, data, current_user.id)

@router.get("/{id}", response_model=SegmentResponse, dependencies=[Depends(require_permission("segments:read"))])
async def get_segment(
    id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return await SegmentService.get_segment(db, id)

@router.patch("/{id}", response_model=SegmentResponse, dependencies=[Depends(require_permission("segments:create"))])
async def update_segment(
    id: UUID,
    data: SegmentUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return await SegmentService.update_segment(db, id, data, current_user.id)

@router.post("/{id}/archive", response_model=SegmentResponse, dependencies=[Depends(require_permission("segments:archive"))])
async def archive_segment(
    id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return await SegmentService.archive_segment(db, id, current_user.id)

@router.post("/{id}/activate", response_model=SegmentResponse, dependencies=[Depends(require_permission("segments:archive"))])
async def activate_segment(
    id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return await SegmentService.activate_segment(db, id, current_user.id)

File: frontend/src/api/segments.ts
import client from './client';
import { Segment, SegmentCreate, SegmentUpdate } from '../types/models';

export const listSegments = async (status: string = 'active', search?: string) => {
  const params: any = { status };
  if (search) params.search = search;
  const response = await client.get<Segment[]>('/api/segments', { params });
  return response.data;
};

export const getSegment = async (id: string) => {
  const response = await client.get<Segment>(`/api/segments/${id}`);
  return response.data;
};

export const createSegment = async (data: SegmentCreate) => {
  const response = await client.post<Segment>('/api/segments', data);
  return response.data;
};

export const updateSegment = async (id: string, data: SegmentUpdate) => {
  const response = await client.patch<Segment>(`/api/segments/${id}`, data);
  return response.data;
};

export const archiveSegment = async (id: string) => {
  const response = await client.post<Segment>(`/api/segments/${id}/archive`);
  return response.data;
};

export const activateSegment = async (id: string) => {
  const response = await client.post<Segment>(`/api/segments/${id}/activate`);
  return response.data;
};
